iothub/service/src/Amqp/ClientWebSocketTransport.cs

```diff
@@ -0,0 +1,412 @@
+        {
+            if (result.CompletedSynchronously)
+            {
+                return;
+            }
+
+            HandleReadComplete(result);
+        }
+
+        private static void HandleReadComplete(IAsyncResult result)
+        {
+            var taskResult = (Task<int>)result;
+            var args = (TransportAsyncCallbackArgs)taskResult.AsyncState;
+
+            ReadTaskDone(taskResult, args);
+            args.CompletedCallback(args);
+        }
+
+        private static bool ReadTaskDone(Task<int> taskResult, TransportAsyncCallbackArgs args)
+        {
+            IAsyncResult result = taskResult;
+            args.BytesTransfered = 0; // reset bytes transferred
+
+            if (taskResult.IsFaulted)
+            {
+                args.Exception = taskResult.Exception;
+                return true;
+            }
+
+            if (taskResult.IsCompleted)
+            {
+                args.BytesTransfered = taskResult.Result;
+                args.CompletedSynchronously = result.CompletedSynchronously;
+                return true;
+            }
+
+            return taskResult.IsCanceled;
+        }
+
+        private static void OnWriteComplete(IAsyncResult result)
+        {
+            if (result.CompletedSynchronously)
+            {
+                return;
+            }
+
+            HandleWriteComplete(result);
+        }
+
+        private static void HandleWriteComplete(IAsyncResult result)
+        {
+            var taskResult = (Task)result;
+            var args = (TransportAsyncCallbackArgs)taskResult.AsyncState;
+            WriteTaskDone(taskResult, args);
+            args.CompletedCallback(args);
+        }
+
+        private static bool WriteTaskDone(Task taskResult, TransportAsyncCallbackArgs args)
+        {
+            args.BytesTransfered = 0; // reset bytes transferred
+            if (taskResult.IsFaulted)
+            {
+                args.Exception = taskResult.Exception;
+                return true;
+            }
+
+            if (taskResult.IsCompleted)
+            {
+                args.BytesTransfered = args.Count;
+                args.CompletedSynchronously = ((IAsyncResult)taskResult).CompletedSynchronously;
+                return true;
+            }
+
+            return taskResult.IsCanceled;
+        }
+
+        private void ThrowIfNotOpen()
+        {
+            WebSocketState webSocketState = _webSocket.State;
+            if (webSocketState == WebSocketState.Open)
+            {
+                return;
+            }
+
+            if (webSocketState == WebSocketState.Aborted
+                || webSocketState == WebSocketState.Closed
+                || webSocketState == WebSocketState.CloseReceived
+                || webSocketState == WebSocketState.CloseSent)
+            {
+                throw new ObjectDisposedException(GetType().Name);
+            }
+
+            throw new AmqpException(AmqpErrorCode.IllegalState, null);
```

Comment:
Will this exception type reach the end user?

Reply:
I'll get rid of this case by collapsing it with the exception above